rcl_logging_implementation/src/functions.cpp

```diff
@@ -0,0 +1,272 @@
+    } catch (const std::exception & e) {
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "failed to resolve symbol '%s' in shared library due to %s",
+        symbol_name.c_str(), e.what());
+    }
+    return nullptr;
+  }
+  return lib->get_symbol(symbol_name);
+}
+
+bool
+load_logging_library()
+{
+  // Already loaded
+  if (g_logging_lib) {
+    return true;
+  }
+
+  // The logic to pick the logging library to load goes as follows:
+  //
+  // 1. If the user specified the library to use via the RCL_LOGGING_IMPLEMENTATION
+  //    environment variable, try to load only that library.
+  // 2. Otherwise, try to load the default logging implementation.
+
+  std::string env_var;
+  try {
+    env_var = rcpputils::get_env_var("RCL_LOGGING_IMPLEMENTATION");
+  } catch (const std::exception & e) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "failed to fetch RCL_LOGGING_IMPLEMENTATION "
+      "from environment due to %s", e.what());
+    return false;
+  }
+
+  // User specified a logging implementation, attempt to load that one and only that one
+  if (!env_var.empty()) {
+    g_logging_lib = attempt_to_load_one_logging_library(env_var);
+  } else {
+    RCUTILS_LOG_DEBUG_NAMED(
+      "rcl_logging_implementation",
+      "RCL_LOGGING_IMPLEMENTATION not set, using default: %s",
+      STRINGIFY(DEFAULT_RCL_LOGGING_IMPLEMENTATION));
+
+    // User didn't specify, so load the default logging implementation
+    g_logging_lib =
+      attempt_to_load_one_logging_library(STRINGIFY(DEFAULT_RCL_LOGGING_IMPLEMENTATION));
+  }
+
+  if (!g_logging_lib) {
+    RCUTILS_SET_ERROR_MSG("failed to load any logging implementations");
+    return false;
+  }
+
+  // Register all function pointers
+  g_initialize_func = reinterpret_cast<rcl_logging_initialize_func_t>(
+    lookup_symbol(g_logging_lib, "rcl_logging_external_initialize"));
+  if (!g_initialize_func) {
+    g_logging_lib.reset();
+    return false;
+  }
+
+  g_shutdown_func = reinterpret_cast<rcl_logging_shutdown_func_t>(
+    lookup_symbol(g_logging_lib, "rcl_logging_external_shutdown"));
+  if (!g_shutdown_func) {
+    g_logging_lib.reset();
+    g_initialize_func = nullptr;
+    return false;
+  }
+
+  g_log_func = reinterpret_cast<rcl_logging_log_func_t>(
+    lookup_symbol(g_logging_lib, "rcl_logging_external_log"));
+  if (!g_log_func) {
+    g_logging_lib.reset();
+    g_initialize_func = nullptr;
+    g_shutdown_func = nullptr;
+    return false;
+  }
+
+  g_set_logger_level_func = reinterpret_cast<rcl_logging_set_logger_level_func_t>(
+    lookup_symbol(g_logging_lib, "rcl_logging_external_set_logger_level"));
+  if (!g_set_logger_level_func) {
+    g_logging_lib.reset();
+    g_initialize_func = nullptr;
+    g_shutdown_func = nullptr;
+    g_log_func = nullptr;
+    return false;
+  }
+
+  RCUTILS_LOG_DEBUG_NAMED(
+    "rcl_logging_implementation",
+    "Successfully registered all function pointers from logging library");
+
+  return true;
```

Comment:
How about handling errors in one position?

```suggestion
  // Register all function pointers
  g_initialize_func = reinterpret_cast<rcl_logging_initialize_func_t>(
    lookup_symbol(g_logging_lib, "rcl_logging_external_initialize"));
  if (!g_initialize_func) {
    goto ERROR;
  }

  g_shutdown_func = reinterpret_cast<rcl_logging_shutdown_func_t>(
    lookup_symbol(g_logging_lib, "rcl_logging_external_shutdown"));
  if (!g_shutdown_func) {
    goto ERROR;
  }

  g_log_func = reinterpret_cast<rcl_logging_log_func_t>(
    lookup_symbol(g_logging_lib, "rcl_logging_external_log"));
  if (!g_log_func) {
    goto ERROR;
  }

  g_set_logger_level_func = reinterpret_cast<rcl_logging_set_logger_level_func_t>(
    lookup_symbol(g_logging_lib, "rcl_logging_external_set_logger_level"));
  if (!g_set_logger_level_func) {
    goto ERROR:
  }

  RCUTILS_LOG_DEBUG_NAMED(
    "rcl_logging_implementation",
    "Successfully registered all function pointers from logging library");

  return true;

ERROR:
   g_logging_lib.reset();
   g_initialize_func = nullptr;
   g_shutdown_func = nullptr;
   g_log_func = nullptr;
   g_set_logger_level_func = nullptr;
   return false;
```

Reply:
i like this, and much better. i will borrow your suggested code, thanks 👍

Reply:
https://github.com/ros2/rcl_logging/pull/135/commits/b925588dbf32b890f1a2d468a7d2c0da02bb8a91 fixes it.